File: submissions/search-a-2d-matrix.go
// https://leetcode.com/problems/search-a-2d-matrix

// (since i previously did search 2d as 1d matrix, do double binary search here)
// binary search the first index, then see whether to search up or search down. 
// Row to search is where matrix[n][0] <= target < matrix[n + 1][0] 
// Once find the row to search, binary search that row to find whether inside

func searchMatrix(matrix [][]int, target int) bool {
    // SC: constant, TC: log(m)*log(n)
    nRows := len(matrix)

    lo := 0
    hi := nRows - 1

    for lo < hi {
        mid := lo + (hi - lo + 1) / 2
        if matrix[mid][0] > target {
            // search up
            hi = mid - 1
        } else {
            // search down
            lo = mid
        }
    }

    return binarySearchRow(matrix[lo], target)

}

func binarySearchRow(row []int, target int) bool {
    // SC: constant. TC: logm
    n := len(row)
    lo := 0
    hi := n - 1
    for lo <= hi {
        mid := lo + (hi - lo) / 2
        if row[mid] == target {
            return true
        } else if row[mid] < target {
            // search right
            lo = mid + 1
        } else {
            // search left
            hi = mid - 1
        }
    }
    
    return false
}

File: submissions/search-a-2d-matrix.py
// https://leetcode.com/problems/search-a-2d-matrix

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        grid = matrix
        nRows = len(grid)
        if nRows == 0:
            return False
        nCols = len(grid[0])

        lo = 0
        hi = nRows * nCols - 1 # last element of the grid
        
        while lo <= hi:
            mid = lo + (hi - lo) // 2
            # find the rowInd and colInd corresponding to mid
            rowInd = mid // nCols # row Index corresponding to mid
            colInd = mid - rowInd * nCols 

            midEle = grid[rowInd][colInd]

            if midEle == target:
                return True
            elif midEle > target: # search left
                hi = mid - 1
            else:
                lo = mid + 1
        return False